### PR TITLE
Rewrote displays type variables to use them as a map

### LINKED ISSUE #26515: More display-* headings - .display-5
### New .display-5 class

When creating website banners I often find myself using one of the `.display-*` headings from Bootstrap. They're clean, simple and look great!

The problem I've encountered is I find myself slightly restricted with the current `.display-*` classes, more specifically with the sizing, I often find myself needing something slightly smaller than `.display-4` and often tend to create my own `.display-5` class, and in the very rare occasion a `.display-6` class.

Could we extend this part and create an additional `.display-65` class?

**Something like:**

```
.display-5 {
    font-size: 2.5rem;
    font-weight: 300;
    line-height: 1.2;
}
```

I'm sure this wouldn't take too long to add?

**This is a mock-up of what the sizing would look like:**

display-5-class

Many Thanks!

:-)

Reference: https://getbootstrap.com/docs/4.1/content/typography/

### PR DESCRIPTION
**Pull request by afanjul**
Rewrited displays type variables to use them as a map so it would be easier to extend and modify.

Fixes #26515.

### CODE REVIEW

**Review by XhmikosR — DISMISSED** [review]
please fix the test failures.

**Comment by afanjul** [comment]
> please fix the test failures.

Could you check now? I tried to just push the fixes to my own branch but the pull request would never get updated (I don't know if its a github bug or what)... so I pushed to my own fork master... according to github message "_Add more commits by pushing to the master branch on afanjul/bootstrap_." 

It's a bit complicated to just update a pull request... :(

**Comment by XhmikosR** [comment]
GitHub could have a slowness issue right now, but usually you just push to your branch.

**Comment by MartijnCuppens** [comment]
Hi @afanjul,

I'm a bit afraid it's not clear what `300` means here for people who don't have a clue the font weight is set:
https://github.com/twbs/bootstrap/blob/3182990fee5aa32b3e4146685e257821c37d9666/scss/_variables.scss#L306-L309

I did a similar PR once, but closed it since I was a bit worried we're mixing up too many different "styles" of configuration: https://github.com/twbs/bootstrap/pull/26533.

**Comment by afanjul** [comment]
> Hi @afanjul,
> 
> I'm a bit afraid it's not clear what `300` means here for people who don't have a clue the font weight is set:
> [bootstrap/scss/_variables.scss](https://github.com/twbs/bootstrap/blob/3182990fee5aa32b3e4146685e257821c37d9666/scss/_variables.scss#L306-L309)
> 
> Lines 306 to 309 in [3182990](/twbs/bootstrap/commit/3182990fee5aa32b3e4146685e257821c37d9666)
> 
>    1:(6rem, 300), 
>    2:(5.5rem, 300), 
>    3:(4.5rem, 300), 
>    4:(3.5rem, 300) 
> I did a similar PR once, but closed it since I was a bit worried we're mixing up too many different "styles" of configuration: #26533.

Another alternative (a worse one from my point of view) is simply split it in two maps:

**In _variables.scss**

```
$display-sizes: (
  1:6rem,
  2:5.5rem,
  3:4.5rem,
  4:3.5rem
);
$display-weights: (
  1:300,
  2:300,
  3:300,
  4:300
);
```

The only "issue" is that you have to use the $index in for loop (and maintain a coherence in the maps):

```
@for $i from 1 through length($display-sizes) {
  $size: nth($display-size, $i);
  $weight: nth($display-weight, $i);
  .display-#{$index} {
    @include font-size($size);
    font-weight: $weight;
    line-height: $display-line-height;
  }
}
```

But the worst method, for me, is the actual one, with separate values/variables for size and weights. It's ugly and not extendible.

**Comment by afanjul** [comment]
Anyway, if you are using Sass as a "web developer", I'm assuming that you can read and interpret docs, comments, functions, etc.  Either way, you will be using compiling bootstrap.css

**Comment by afanjul** [comment]
For example, what if I want to add another "display-5" in the actual bootstrap version? 
I can add a new $spacers or $sizes element, but you don't let me do the same with "$displays"? A bit odd..

File: scss/_variables.scss [hunk 1]
```diff
@@ -301,15 +301,13 @@ $headings-font-weight:        500 !default;
...
-$display3-size:               4.5rem !default;
-$display4-size:               3.5rem !default;
-
-$display1-weight:             300 !default;
-$display2-weight:             300 !default;
-$display3-weight:             300 !default;
-$display4-weight:             300 !default;
+//display types as font-size and font-weight
+$displays: (
+  1:(6rem, 300),
```
Comment (mdo) [inline scss/_variables.scss]: Let's use nested maps with explicit key-value pairs instead of this setup that requires `nth` value navigation.

**Review by mdo — CHANGES_REQUESTED** [review]

**Comment by mdo** [comment]
> I did a similar PR once, but closed it since I was a bit worried we're mixing up too many different "styles" of configuration: #26533.

Reading that thread again @MartijnCuppens, it looks like I punted on it for redoing bigger things in v5. I can't recall exactly what though, but perhaps it was related to your RFS work?

File: scss/_type.scss [hunk 2]
```diff
@@ -26,10 +26,10 @@ h6, .h6 { @include font-size($h6-font-size); }
 }
 
 // Type display classes
-@each $index, $values in $displays {
+@each $index, $display-type in $display-types {
   .display-#{$index} {
-    @include font-size(#{nth($values, 1)});
-    font-weight: #{nth($values, 2)};
+    @include font-size(map-get($display-type, font-size));
+    font-weight: map-get($display-type, font-weight);
```
Comment (afanjul) [inline scss/_type.scss]: I'm getting an "Expected numeric font-weight notation" css-lint error in the map-get function, but all seems right and the compiling output is also right... so I don't know...

Comment (MartijnCuppens) [reply scss/_type.scss]: You can disable this with `// stylelint-disable-next-line font-weight-notation`. Stylelint doesn't know what the `map-get` result will be, so you need to explicitly disable the rule on that line.

**Comment by afanjul** [comment]
Now I think that the source code is complete and prettier... Using a map for $display-types, with key value explicit names. So now is pretty easy to modify or extend by the people... Just be aware of the new variable called $display-types for docs, and so...

File: scss/_variables.scss (hunk 1 above)
Comment (afanjul) [reply scss/_variables.scss]: Now using nested maps with explicit key-value pairs

**Comment by XhmikosR** [comment]
This needs a rebase + squash

**Comment by mdo** [comment]
Replaced by #30568.

**Pull request closed without merging**